src/Services/CDADocumentService.php

```diff
@@ -0,0 +1,177 @@
+<?php
+
+/**
+ * Document Service for CCDA
+ *
+ * @package   OpenEMR
+ * @link      https://www.open-emr.org
+ * @author    Jerry Padgett <sjpadgett@gmail.com>
+ * @copyright Copyright (c) 2021 Jerry Padgett <sjpadgett@gmail.com>
+ * @license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
+ */
+
+namespace OpenEMR\Services;
+
+use CouchDB;
+use OpenEMR\Common\Crypto\CryptoGen;
+use OpenEMR\Common\Uuid\UuidRegistry;
+use Symfony\Component\HttpClient\HttpClient;
+
+/**
+ * Class CDADocumentService
+ *
+ * @package OpenEMR\Services
+ *
+ * See interface/modules/zend_modules/module/Carecoordination/src/Carecoordination/Controller/EncounterccdadispatchController.php
+ * indexAction() and interface/modules/zend_modules/public/index.php
+ */
+class CDADocumentService extends BaseService
+{
+    const TABLE_NAME = "ccda";
+    protected $serverUrl;
+
+    public function __construct()
+    {
+        parent::__construct(self::TABLE_NAME);
+        UuidRegistry::createMissingUuidsForTables([self::TABLE_NAME]);
+        $this->serverUrl = $GLOBALS['qualified_site_addr'];
+    }
+
+    public function getLastCdaMeta($pid)
+    {
+        $query = "SELECT cc.uuid, cc.date, pd.fname, pd.lname, pd.pid FROM ccda AS cc
+		    LEFT JOIN patient_data AS pd ON pd.pid=cc.pid
+		    WHERE cc.pid = ?
+		    ORDER BY cc.id DESC LIMIT 1";
+
+        return sqlQuery($query, array($pid));
+    }
+
+    public function getFile($id)
+    {
+        $query = "select couch_docid, couch_revid, ccda_data, encrypted from ccda where uuid=?";
+        $row = sqlQuery($query, array($id));
+        $content = '';
+        if (!empty($row)) {
+            if (!empty($row['couch_docid'])) {
+                $couch = new CouchDB();
+                $resp = $couch->retrieve_doc($row['couch_docid']);
+                if ($row['encrypted']) {
+                    $cryptoGen = new CryptoGen();
+                    $content = $cryptoGen->decryptStandard($resp->data, null, 'database');
+                } else {
+                    $content = base64_decode($resp->data);
+                }
+            } elseif (!empty($row['ccda_data'])) {
+                $fccda = fopen($row['ccda_data'], "r");
+                if ($row['encrypted']) {
+                    $cryptoGen = new CryptoGen();
+                    $content = $cryptoGen->decryptStandard(fread($fccda, filesize($row['ccda_data'])), null, 'database');
+                } else {
+                    $content = fread($fccda, filesize($row['ccda_data']));
+                }
+                fclose($fccda);
+            }
+        }
+
+        return $content;
+    }
+
+    public function generateCCDHtml($pid): string
```

Comment:
would it make sense to bypass the need to send all these "outside" requests? I'm happy to look into how to go through encounterccdadispatch without needing it to be external (ie. need to send a session_id).

Reply:
I think it's fine as is. It's not like the id is a secret(in cookie) and is no different than globals opening the session which it still does anyway to do auth. Besides, this is a normal way to handle the sessions in api calls via http.


Up to you but i'd like to avoid it being too cumbersome to call.